renderer/backend/DrawBatchExecutor.cpp

```diff
@@ -0,0 +1,237 @@
+ on the Doom 3 GPL Source Code as published in 2011.
+
+ The Dark Mod Source Code is free software: you can redistribute it
+ and/or modify it under the terms of the GNU General Public License as
+ published by the Free Software Foundation, either version 3 of the License,
+ or (at your option) any later version. For details, see LICENSE.TXT.
+
+ Project: The Dark Mod (http://www.thedarkmod.com/)
+
+******************************************************************************/
+#include "precompiled.h"
+#include "DrawBatchExecutor.h"
+
+#include "../glsl.h"
+#include "../qgl.h"
+
+const uint DrawBatchExecutor::DEFAULT_UBO_INDEX;
+const uint DrawBatchExecutor::MAX_SHADER_PARAMS_SIZE;
+
+namespace {
+	const uint MAX_DRAWS_PER_FRAME = 8192;
+
+	struct DrawElementsIndirectCommand {
+		uint count;
+		uint instanceCount;
+		uint firstIndex;
+		uint baseVertex;
+		uint baseInstance;
+	};
+}
+
+idCVarBool r_useMultiDrawIndirect("r_useMultiDrawIndirect", "1", CVAR_RENDERER|CVAR_BOOL, "Batch draw calls in multidraw commands if available");
+
+void DrawBatchExecutor::Init() {
+	GLint uboAlignment;
+	qglGetIntegerv( GL_UNIFORM_BUFFER_OFFSET_ALIGNMENT, &uboAlignment );
+	qglGetIntegerv(GL_MAX_UNIFORM_BLOCK_SIZE, &maxUniformBlockSize);
+
+	uint shaderParamsBufferSize = MAX_DRAWS_PER_FRAME * MAX_SHADER_PARAMS_SIZE;
+	shaderParamsBuffer.Init( GL_UNIFORM_BUFFER, shaderParamsBufferSize, uboAlignment );
+
+	if (GLAD_GL_ARB_multi_draw_indirect) {
+		InitDrawIdBuffer();	
+		uint drawCommandBufferSize = MAX_DRAWS_PER_FRAME * sizeof(DrawElementsIndirectCommand);
+		drawCommandBuffer.Init( GL_DRAW_INDIRECT_BUFFER, drawCommandBufferSize, 16 );
+		drawCommandBuffer.Bind();
+	}
+}
+
+void DrawBatchExecutor::Destroy() {
+	shaderParamsBuffer.Destroy();
+	qglDeleteBuffers(1, &drawIdBuffer);
+	drawIdBuffer = 0;
+	drawCommandBuffer.Destroy();
+}
+
+void DrawBatchExecutor::ExecuteDrawVertBatch( int numDrawSurfs, GLuint uboIndex ) {
+	assert( numDrawSurfs <= maxBatchSize );	
+	maxBatchSize = 0;
+
+	byte *shaderParamsContents = shaderParamsBuffer.CurrentWriteLocation();
+	uint shaderParamsCommitSize = numDrawSurfs * shaderParamsSize;
+	shaderParamsBuffer.Commit( shaderParamsCommitSize );
+	shaderParamsBuffer.BindRange( uboIndex, shaderParamsContents, shaderParamsCommitSize );
+
+	vertexCache.BindVertex();
+	vertexCache.BindIndex();
+	if ( ShouldUseMultiDraw() ) {
+		DrawVertsMultiDraw( numDrawSurfs );
+	} else {
+		DrawVertsSingleDraws( numDrawSurfs );
+	}
+
+	if ( r_showPrimitives.GetBool() && !backEnd.viewDef->IsLightGem() && backEnd.viewDef->viewEntitys ) {
+		for ( int i = 0; i < numDrawSurfs; ++i ) {
+			const drawSurf_t *surf = drawSurfs[i];
+			backEnd.pc.c_drawIndexes += surf->numIndexes;
+			if ( surf->frontendGeo ) {
+				backEnd.pc.c_drawVertexes += surf->frontendGeo->numVerts;
+			}
+			backEnd.pc.c_vboIndexes += surf->numIndexes;
+		}
+		backEnd.pc.c_drawElements += ShouldUseMultiDraw() ? 1 : numDrawSurfs;
```

Comment:
Perhaps we should add a counter for the number of batches sent to driver?

There are so many conditions for state changes which separate stream of surfaces into batches, that it is non-trivial to understand how a change would affect the number of batches.

Reply:
For now, I've reverted it to always count the number of draw surfaces, which is consistent with what it's been doing so far. Adding another counter is a bit more effort, and it's probably irrelevant for most people. If you really need to see/measure the actual draw calls issued, tools like RenderDoc or nSight will give you a better insight, anyway.

Reply:
I think many mappers look at draw calls count using r_showPrimitives or somewthing like that. I guess they would be disappointed if you radically change this metric in a way like you did =)

Regarding new counter, I think the question is: can it be useful on any other machine except for developers' machines? Can you see the number easily in RenderDoc?

Reply:
I'm not sure you can see the number easily in RenderDoc. In nSight, you can profile a frame, which will tell you the number of issued actual draw calls.

But I think the entire point of this is that the number of draw calls should be reduced to the point that the exact number of either the draw calls or the underlying drawn objects should not matter too much anymore, so mappers can concentrate on the number of tris and vertices, instead. So I'm not sure the extra counter is all that useful. So I'd suggest we add it if someone does find a use for it :)